ownCloud/Client/Actions/Actions+Extensions/DuplicateAction.swift

```diff
@@ -34,37 +37,126 @@ class DuplicateAction : Action {
 
 	// MARK: - Action implementation
 	override func run() {
-		guard context.items.count > 0, let core = self.core, let item = context.items.first, let itemName = item.name else {
+		guard context.items.count > 0 else {
 			completed(with: NSError(ocError: .itemNotFound))
 			return
 		}
+		remainingItems.append(contentsOf: context.items)
+		proceed()
+	}
 
-		context.items.forEach({ (item) in
-			if let rootItem = item.parentItem(from: core) {
+	func proceed() {
+		guard let item = remainingItems.first, let baseName = item.baseName, let core = self.core, let itemPath = item.path, let rootItemPath = itemPath.parentPath() else {
+			completed(with: NSError(ocError: .itemNotFound))
+			return
+		}
 
-				var name: String = "\(itemName) copy"
+		let pattern = " \(localizedCopy)[ ]?[0-9]{0,}$"
+		let searchName = baseName.replacingOccurrences(for: pattern)
 
-				if let itemFileExtension = item.fileExtension, let baseName = item.baseName {
-					var fileExtension = itemFileExtension
+		query = OCQuery(forPath: rootItemPath)
+		query?.delegate = self
+		query?.queryItem = item
 
-					if fileExtension != "" {
-						fileExtension = ".\(fileExtension)"
-					}
+		let filterHandler: OCQueryFilterHandler = { (_, _, item) -> Bool in
+			if let itemName = item?.baseName {
+				return itemName.hasPrefix(searchName)
+			}
+			return false
+		}
+		query?.addFilter(OCQueryFilter.init(handler: filterHandler), withIdentifier: "rename-file")
 
-					name = "\(baseName) copy\(fileExtension)"
+		core.start(query!)
+	}
+
+	// MARK: - Query handling
+	func query(_ query: OCQuery, failedWithError error: Error) {
+		completed(with: NSError(ocError: .itemNotFound))
+	}
+
+	func queryHasChangesAvailable(_ query: OCQuery) {
+		query.requestChangeSet(withFlags: OCQueryChangeSetRequestFlag(rawValue: 0)) { (query, changeSet) in
+
+			guard let queryItem = query.queryItem else { return }
+
+			switch query.state {
+			case .contentsFromCache, .waitingForServerReply:
+				break
+			case .idle:
+				self.core?.stop(query)
+
+				let items = changeSet?.queryResult ?? []
+				guard items.first != nil else {
```

Comment:
@hosy why do we have extra handling code for the first item? And then we have a loop for all of them? Can't we have a generic loop iterating over all of them?

Reply:
Maybe `items.first` is wrong. But if you look at the code, I'm working with delegate methods for OCQuery. If you put it into a loop it would not work as expected, because if you duplicate two files with the same name, but with ending ` copy x`, both files would get the same name. So you have to copy the first item, before you can name the second item.

Reply:
Just checked, that `items.first` is not wrong, because if it not exists, there is an important fallback code.

Reply:
@hosy sorry, sometimes I am just questioning the code which is a bit hard to read at a first glance)